Prevent duplicate upgrades (plugs) to GDI Upgrade Center (Pluggable Requirements)

atlimit8 opened the pull request:
While I was looking at the `BooleanExpression`/condition code,  I noticed that the **GDI Upgrade Center** accepts two of the same type of upgrade/plug. This forces the **GDI Upgrade Center** to accept only one **Seeker Control** and only one **Ion Cannon Uplink**.

This PR adds the ability to use `ConditionExpression`s for allowing a plug type instead of checking for an existing plug.

## Tests
### Component Tower armament replacement test
This allows replacing the armament of a Component Tower:
```diff
diff --git a/mods/ts/rules/gdi-support.yaml b/mods/ts/rules/gdi-support.yaml
index 31f5c8723e..8c5977ff73 100644
--- a/mods/ts/rules/gdi-support.yaml
+++ b/mods/ts/rules/gdi-support.yaml
@@ -130,6 +130,10 @@ GACTWR:
 			tower.vulcan: tower.vulcan
 			tower.rocket: tower.rocket
 			tower.sam: tower.sam
+		Requirements:
+			tower.vulcan: !tower.vulcan
+			tower.rocket: !tower.rocket
+			tower.sam: !tower.sam
 	ProvidesPrerequisite@buildingname:
 	SelectionDecorations:
 		VisualBounds: 48, 48, 0, -12
```
### Replace with like upgrade for GDI Upgrade Center Test
This tests that the support powers don't reset their timers.
```diff
diff --git a/mods/ts/rules/gdi-structures.yaml b/mods/ts/rules/gdi-structures.yaml
index a27a7c6d71..10370c7c00 100644
--- a/mods/ts/rules/gdi-structures.yaml
+++ b/mods/ts/rules/gdi-structures.yaml
@@ -445,8 +445,8 @@ GAPLUG:
 			plug.ioncannon: plug.ioncannona
 			plug.hunterseeker: plug.hunterseekera
 		Requirements:
-			plug.ioncannon: !plug.ioncannonb && !plug.ioncannona && !plug.hunterseekera
-			plug.hunterseeker: !plug.hunterseekerb && !plug.ioncannona && !plug.hunterseekera
+			plug.ioncannon: !plug.ioncannonb
+			plug.hunterseeker: !plug.hunterseekerb
 	WithIdleOverlay@ioncannona:
 		RequiresCondition: plug.ioncannona
 		Sequence: idle-ioncannona
@@ -459,8 +459,8 @@ GAPLUG:
 			plug.ioncannon: plug.ioncannonb
 			plug.hunterseeker: plug.hunterseekerb
 		Requirements:
-			plug.ioncannon: !plug.ioncannona && !plug.ioncannonb && !plug.hunterseekerb
-			plug.hunterseeker: !plug.hunterseekera && !plug.ioncannonb && !plug.hunterseekerb
+			plug.ioncannon: !plug.ioncannona
+			plug.hunterseeker: !plug.hunterseekera
 	WithIdleOverlay@ioncannonb:
 		RequiresCondition: plug.ioncannonb
 		Sequence: idle-ioncannonb
```

atlimit8 committed:
Added ConditionExpression Requirements for plugs to Pluggable

atlimit8 committed:
Prevent duplicate upgrades (plugs) to GDI Upgrade Center

GraionDilach commented:
That's due to mimicng legacy behaviour, regardless of how stupid it might sound like.

pchote commented on OpenRA.Mods.Common/Traits/Pluggable.cs:
```diff
@@ -23,19 +25,29 @@ public class PluggableInfo : ITraitInfo, UsesInit<PlugsInit>
+		[ConsumedConditionReference]
+		public IEnumerable<string> ConsumedConditions
+		{
+			get { return Requirements.Values.SelectMany(r => r.Variables.Except(Conditions.Values)); }
```
Why are you filtering out the granted conditions?  This doesn't look right.

atlimit8 replied:
You're right, that is for the linter. I'll have to split this into `ConsumedConditions` and `LinterConsumedConditions` or is it valid to consider the same trait as a valid consumer of a condition?.

pchote replied:
It's valid for the same trait to both grant and consume the same condition.  If said trait was the only granter and consumer (not useful in this trait, but could be in others) then the linter would fail if it weren't listed in both.

atlimit8 commented:
The linter and `ConditionManager` new use separate properties.

atlimit8 commented:
Now both the linter and `ConditionManager` consider `Pluggable` a consumer of conditions granted by itself.

atlimit8 commented:
Fixed copy and paste description.

atlimit8 committed:
PluggableInfo.Requirements entry overrides in-use check now to allow swapping.

atlimit8 commented:
Made changes and updated description.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs:
```diff
@@ -191,6 +222,54 @@ public int GrantCondition(Actor self, string condition, bool external = false, i
+		/// <param name="newCondition">Replacement condition to grant.</param>
+		/// <param name="external">Validate against the external condition whitelist.</param>
+		/// <param name="duration">Automatically revoke condition after this delay if non-zero.</param>
+		public int ReplaceCondition(Actor self, int oldToken, string newCondition, bool external = false, int duration = 0)
```
I'd prefer that we didn't expose this, it just adds complexit.
If a trait wants to change what it grants it should explicitly revoke the old one and grant the new one.

atlimit8 commented:
Removed `ConditionManager.ReplaceCondition`.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Pluggable.cs:
```diff
@@ -67,6 +90,8 @@ public void EnablePlug(Actor self, string type)
 				return;
 
+			if (conditionToken != ConditionManager.InvalidConditionToken)
+				conditionManager.RevokeCondition(self, conditionToken);
```
Newline after this, please.

atlimit8 replied:
Done.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Pluggable.cs:
```diff
@@ -58,7 +77,11 @@ public void Created(Actor self)
+			if (!Info.Conditions.ContainsKey(type))
+				return false;
+			if (!Info.Requirements.ContainsKey(type))
+				return active == null;
```
Newlines after the returns.

pchote commented on OpenRA.Mods.Common/Traits/Pluggable.cs:
```diff
@@ -46,6 +58,13 @@ public Pluggable(ActorInitializer init, PluggableInfo info)
+			{
+				plugTypesAvailability = new Dictionary<string, bool>();
+				foreach (var plug in info.Requirements)
+					plugTypesAvailability.Add(plug.Key, true);
```
It would be better to use `plugTypesAvailability[plug.Key] = true` here to avoid a crash if someone defines duplicate requirements.

pchote reviewed (approved):
Looks good to me!  :+1: after fixing @Phrohdoh's style nits, and a couple more minor requests below:

reaperrr commented on OpenRA.Mods.Common/Traits/Pluggable.cs:
```diff
@@ -20,22 +22,36 @@ public class PluggableInfo : ITraitInfo, UsesInit<PlugsInit>
 
+		[Desc("Requirements for accepting a plug type.",
+			"Key is the plug type that the requirements applies to.",
+			"Value is the condition expression definining the requirements to place the plug.")]
```
Typo: definining -> defining

reaperrr commented:
👍 after that typo is fixed.

atlimit8 commented:
Fixed typo and rebased on `ConditionExpression` (minor `bool` => `int`, `BooleanExpression` => `ConditionExpression`, `.Evaluate(conditions)` => `.Evaluate(conditions) != 0` changes).
